synapse/visibility.py

```diff
@@ -102,165 +105,369 @@ async def filter_events_for_client(
+        state: The state at the event, unless its an outlier
+        sender_erased: Whether the event sender has been marked as "erased"
+
+    Returns:
+        None if the user cannot see this event at all
+
+        a redacted copy of the event if they can only see a redacted
+        version
+
+        the original event if they can see it as normal.
+    """
+    # Only run some checks if these events aren't about to be sent to clients. This is
+    # because, if this is not the case, we're probably only checking if the users can
+    # see events in the room at that point in the DAG, and that shouldn't be decided
+    # on those checks.
+    if filter_send_to_client:
+        if not _check_filter_send_to_client(
+            event, clock, retention_policy, sender_ignored
+        ):
             return None
 
-        # the visibility is either shared or world_readable, and the user was
-        # not a member at the time. We allow it, provided the original sender
-        # has not requested their data to be erased, in which case, we return
-        # a redacted version.
-        if erased_senders[event.sender]:
-            return prune_event(event)
+    if event.event_id in always_include_ids:
+        return event
+
+    # we need to handle outliers separately, since we don't have the room state.
+    if event.internal_metadata.outlier:
+        # Normally these can't be seen by clients, but we make an exception for
+        # for out-of-band membership events (eg, incoming invites, or rejections of
+        # said invite) for the user themselves.
+        if event.type == EventTypes.Member and event.state_key == user_id:
+            logger.debug("Returning out-of-band-membership event %s", event)
+            return event
+
+        return None
+
+    if state is None:
+        raise Exception("Missing state for non-outlier event")
+
+    # If the sender has been erased we must only return the redacted form.
+    if sender_erased:
+        event = prune_event(event)
+
+    # get the room_visibility at the time of the event.
+    visibility = get_effective_room_visibility_from_state(state)
 
+    # Check if the room has lax history visibility, allowing us to skip
+    # membership checks.
+    if _check_history_visibility(event, visibility, is_peeking):
         return event
 
-    # Check each event: gives an iterable of None or (a potentially modified)
-    # EventBase.
-    filtered_events = map(allowed, events)
+    if _check_membership(user_id, event, visibility, state, is_peeking):
+        return event
 
-    # Turn it into a list and remove None entries before returning.
-    return [ev for ev in filtered_events if ev]
+    return None
+
+
+def _check_membership(
+    user_id: str,
+    event: EventBase,
+    visibility: str,
+    state: StateMap[EventBase],
+    is_peeking: bool,
+) -> bool:
+    """Check whether the user can see the event due to their membership
+
+    Returns:
+        True if they can, False if they can't
```

Comment:
There are now three `_check_*` functions with similar-but-different descriptions of their return values.

- _check_membership: True if user can definitely see the event, False if maybe not.
- _check_filter_to_send_to_client: True if might be allowed to be sent to clients, False if definitely not.
- _check_history_visibility: True if user can definitely see the event, False if maybe not.

This seems like a point of confusion to me. Could these functions' return values be modelled with an enum?